### PR TITLE
DRAFT: Add custom variable updater.

### PR DESCRIPTION
**Pull request by cantonios**
Allows customization for how variables are updated by the optimizer. The base optimizer simply defers to the update handler to do the update, allowing full customization.

Can replace the existing `overwrite_with_gradient` attribute on variables, which currently is very application-specific.

Eliminates creation of optimizer variables that have custom updaters (including `overwrite_with_gradient`), since those variables are never used and may be wasteful.

This is an alternative to #21196.  It would allow us to add special-handling for large embedding tables, where we do not want to pass around large gradients for tables that might span multiple devices.  Instead the tables are updated in-place using a custom update rule.

### CODE REVIEW

**Comment by cantonios** [comment]
@hertschuh @fchollet

Potential replacement for #21196.  With this approach, layering is mostly preserved (although in our case the custom updater would still actually contain an optimizer - but at least it doesn't need to in the general case).

This change can also allow us to remove the `overwrite_with_gradient` attribute entirely, instead using a custom updater.  This attribute is very specific to scale factors in fp8 quantization.  I otherwise potentially _could_ have used this attribute myself, if it weren't for how it handles gradient accumulation steps where it takes the max.

File: keras/src/optimizers/optimizer.py [hunk 1]
```diff
@@ -23,5 +24,90 @@ class Optimizer(BackendOptimizer, base_optimizer.BaseOptimizer):
     pass
 
 
+@keras_export("keras.optimizers.VariableUpdater")
+class VariableUpdater:
+    """Allows special handling of variable updates."""
```
Comment (fchollet) [inline keras/src/optimizers/optimizer.py:29]: If this is intended to be public in the API then we should provide an explanation and a usage example in the docstring

**Review by fchollet — COMMENTED** [review]
This approach looks good to me!

File: keras/src/optimizers/optimizer.py (hunk 1 above)
Comment (cantonios) [reply keras/src/optimizers/optimizer.py:29]: Great!  Yes, I'd like it to be public so I can inherit from it without needing to import internal keras APIs.

Is this location (within `optimizer.py`) okay, or would you prefer it somewhere else, (e.g. in its own file)?
